SonarQube.Common/SarifUtilities.cs

```diff
@@ -0,0 +1,91 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="SarifUtilities.cs" company="SonarSource SA and Microsoft Corporation">
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace SonarQube.Common
+{
+    public class SarifUtilities
+    {
+        /// <summary>
+        /// Returns true if the given SARIF came from the VS 2015 RTM Roslyn, which does not provide correct output.
+        /// </summary>
+        public static bool IsSarifFromUnsupportedCompiler(string input)
+        {
+            // low risk of false positives / false negatives
+            if (input.Contains(@"""toolName"": ""Microsoft (R) Visual C# Compiler""")
+                && input.Contains(@"""productVersion"": ""1.0.0"""))
+            {
+                return true;
+            }
+
+            // all other cases
+            return false;
+        }
+
+        /// <summary>
+        /// Used to correct invalid SARIF generated by VS 2015 RTM Roslyn. 
+        /// Replaces all slashes in any line beginning with "uri": with double slashes.
+        /// Applying this to already-valid SARIF will result in over-escaping.
+        /// </summary>
+        public static bool FixImproperlyEscapedSarif(string input, out string output)
+        {
+            string[] inputLines = input.Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
+            bool changeApplied = false;
+
+            if (IsSarifFromUnsupportedCompiler(input))
+            {
+                for (int i = 0; i < inputLines.Length; i++)
+                {
+                    string line = inputLines[i];
+                    if (line.Contains(@"""uri"": ")
+                        || line.Contains(@"""shortMessage"": ")
+                        || line.Contains(@"""fullMessage"": ")
```

Comment:
The scenario is possible but judged to be unlikely enough that it's not worth the technical complexity of writing a multiline solution for.